loom/common/core/telemetry.cpp

```diff
@@ -0,0 +1,203 @@
+#include "loom/common/core/telemetry.h"
+
+#include "loom/common/assets/assets.h"
+#include "loom/common/core/log.h"
+#include "loom/common/core/performance.h"
+#include "loom/common/utils/fourcc.h"
+
+lmDefineLogGroup(gTelemetryLogGroup, "lt", true, LoomLogInfo)
+
+bool Telemetry::enabled = false;
+bool Telemetry::pendingEnabled = false;
+
+utByteArray Telemetry::sendBuffer;
+
+int Telemetry::tickId = 0;
+
+TableValues<TickMetricValue> Telemetry::tickValues;
+
+loom_precision_timer_t Telemetry::tickTimer = loom_startTimer();
+utArray<utString> Telemetry::tickTimerStack;
+TableValues<TickMetricRange> Telemetry::tickRanges;
+
+// Initialize specialized constants for every type used
+// TickMetricValue
+template<> const TableType TableValues<TickMetricValue>::type = 1;
+template<> const int TableValues<TickMetricValue>::packedItemSize = 4 + 8;
+// TickMetricRange
+template<> const TableType TableValues<TickMetricRange>::type = 2;
+template<> const int TableValues<TickMetricRange>::packedItemSize = 4 + 4 * 4 + 2 * 8;
+
+
+void Telemetry::enable()
+{
+    pendingEnabled = true;
+}
+
+void Telemetry::disable()
+{
+    pendingEnabled = false;
+}
+
+void Telemetry::beginTick()
+{
+    if (enabled != pendingEnabled) {
+        enabled = pendingEnabled;
+        lmLog(gTelemetryLogGroup, "Telemetry %s", enabled ? "enabled" : "disabled");
+    }
+    if (!enabled) return;
+
+    tickValues.reset();
+    tickRanges.reset();
+    tickTimerStack.clear();
+
+    loom_resetTimer(tickTimer);
+
+    // Add the tick id as a default tick value
+    setTickValue("tick.id", tickId);
+}
+
+void Telemetry::endTick()
+{
+    if (!enabled) return;
+
+    lmAssert(tickTimerStack.size() == 0, "Tick timer end call missing");
+
+    // Customized asset protocol message (3 ints + tables)
+    int sendSize = 3 * 4 + tickValues.size + tickRanges.size;
+    sendBuffer.resize(sendSize);
+    sendBuffer.setPosition(0);
+    sendBuffer.writeInt(sendSize);
+    sendBuffer.writeInt(0xDEADBEEF);
```

Comment:
Yes, it's part of the asset protocol, usually the code just has the standard header copied, but I could clean it up in a separate PR.